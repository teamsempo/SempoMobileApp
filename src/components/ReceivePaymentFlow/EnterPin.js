import React from 'react'
import { StyleSheet, Text, View, ActivityIndicator, StatusBar } from 'react-native'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import {connect} from "react-redux";
import { strings } from '../../../locales/i18n';

import TransferAmount from './TransferAmountDisplay'
import Keypad from '../Keypad'
import Styles from "../../Styles";
import { createTransferRequest } from "../../reducers/creditTransferReducer";

const mapStateToProps = (state) => {
    return {
        newTransferStatus: state.creditTransfers.createStatus,
        transferData: state.creditTransfers.transferData
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        createTransferRequest: (body) => dispatch(createTransferRequest({body}))
    };
};

class EnterPin extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            pin: '',
            missing_pin: false,
            incorrect_pin: false,
        };
        this._handleChange = this._handleChange.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.setState({incorrect_pin: (nextProps.newTransferStatus.error)});
    }

    _handleChange(pin) {
        this.setState(pin);
    }

    _handleQrTransfer() {
        if (this.state.pin.length < 4) {
            this.setState({missing_pin: true});
            return
        }

        this.setState({missing_pin: false});
        this.props.createTransferRequest({
            'public_identifier': this.props.transferData.public_identifier,
            'nfc_id': null,
            'transfer_amount': this.props.transferData.transfer_amount,
            'transfer_use': this.props.transferData.transfer_use,
            'pin': this.state.pin
        })
    }

    _onConfirmTransfer() {
        this._handleQrTransfer()
    }

    render() {
        if (this.props.newTransferStatus.isRequesting) {
            return (
                <View style={styles.rootContainer}>
                    <View style={styles.loadingContainer}>
                        <ActivityIndicator />
                        <StatusBar barStyle="default" />
                    </View>
                </View>
            )
        } else {

            const inputButtons = [
                [1, 2, 3],
                [4, 5, 6],
                [7, 8, 9],
                [
                    {value: 'B', display: <Icon name="backspace" size={20} color="#3D454C"/>},
                    0,
                    {value: 'S', display: <Icon name="check-circle" size={20} color="#3D454C"/>}
                ]
            ];


            if (this.props.newTransferStatus.error) {
                var inner_text = (
                    <Text style={{marginTop: 10, marginLeft: 20, marginRight: 20, textAlign: 'center', color: 'red'}}>
                        {this.props.newTransferStatus.error.message}
                    </Text>
                )
            } else if (this.state.missing_pin) {
                inner_text = (
                    <Text style={{marginTop: 10, marginLeft: 20, marginRight: 20, textAlign: 'center', color: 'red'}}>
                        {strings('EnterPin.MissingPin')}
                    </Text>
                )
            } else {
                inner_text = (
                    <Text style={{marginTop: 10, marginLeft: 20, marginRight: 20, textAlign: 'center'}}>
                        {strings('EnterPin.EnterPinText')}
                    </Text>
                )
            }

            return (
                <View style={styles.rootContainer}>
                    <View style={styles.promptContainer}>

                        <TransferAmount/>

                        { inner_text }

                    </View>

                    <View style={styles.pinDisplayContainer}>

                        <Text style={Styles.displayText}>
                            {"* ".repeat(Math.max(this.state.pin.length - 1,0)) + this.state.pin.slice(-1)}
                        </Text>

                    </View>

                    <Keypad value={this.state.pin} inputButtons={inputButtons} onChange={(pin) => this.setState({pin})} onSubmit={() => this._onConfirmTransfer()}/>
                </View>
            )
        }
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(EnterPin);


const styles = StyleSheet.create({
    rootContainer: {
        flex: 1,
        backgroundColor: '#FFF',
    },
    mainContainer: {
        flex: 11,
        justifyContent: 'center',
        alignItems: 'center',
    },
    promptContainer: {
        flex: 4,
        justifyContent: 'center',
        alignItems: 'center',
    },
    pinDisplayContainer: {
        flex: 2,
        justifyContent: 'center',
        alignItems: 'center',
    },
    loadingContainer: {
        flex: 10,
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: '#FFF',
    }
});