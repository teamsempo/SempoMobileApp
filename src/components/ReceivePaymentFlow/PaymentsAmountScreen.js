import React from 'react'
import {StyleSheet, Text, View, Image, Linking, TouchableNativeFeedback, Dimensions} from 'react-native'
import {connect} from "react-redux";

import {isRTL, strings} from '../../../locales/i18n';

import Styles from '../../Styles';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import PaymentsKeypad from '../PaymentsKeypad'

import { updateTransferData } from "../../reducers/creditTransferReducer.js"
import PaymentModeSwitcher from "../PaymentModeSwitcher";
import ExitToHome from "../ExitToHome"
import {removeNFCCardError} from "../../reducers/creditTransferReducer";
import ShareWallet from "../ShareWallet";

const mapStateToProps = (state) => {
    return {
        NFCCheck: state.NFCCheck,
        NFCTransfer: state.NFCTransfer,
        transferData: state.creditTransfers.transferData,
        login: state.login,
        transferAccounts: state.transferAccounts,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        updateTransferData: (payload) => dispatch(updateTransferData(payload)),
        removeNFCCardError: () => dispatch(removeNFCCardError())
    };
};


class PaymentsAmountScreen extends React.Component {

    handle_keypad_change(amount) {
        let key = this.makeKey();
        console.log('new key:', key)

        this.props.updateTransferData({
            transfer_amount: parseInt(amount),
            transfer_random_key: key,
        });

        this.props.removeNFCCardError()

    }

    componentWillUnmount() {
        this.props.updateTransferData({transfer_amount: null})
    }

    makeKey() {
        let text = "";
        const possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

        for (var i = 0; i < 4; i++)
            text += possible.charAt(Math.floor(Math.random() * possible.length));


        return text;
    }

    nextPaymentScreen() {
        if (this.props.transferData.transfer_amount <= 0) {
            return
        } else if (this.props.login.transferUsages.length === 0) {
            this.props.navigation.navigate('PaymentOption', { title: strings('GoBackNav.DefaultTitleText')} )
        } else {
            this.props.navigation.navigate('PaymentUse', { title: strings('GoBackNav.DefaultTitleText')} )
        }
    };

    openRefund() {
        if (this.props.transferData.transfer_amount > 0) {
            this.props.navigation.navigate('Camera', { title: strings('GoBackNav.DefaultTitleText')} )
        } else {
            return
        }
    };


    render() {
        const transferAccount = this.props.login.transferAccountId && this.props.transferAccounts.byId[this.props.login.transferAccountId];
        const displayDecimals = this.props.login.displayDecimals;
        const transferAmount = (this.props.transferData.transfer_amount);
        const isCharging = !this.props.transferData.is_sending;
        const transferDirection = isCharging? 'charge':'refund';

        return (
            <View style={Styles.rootContainer}>

                <View style={[{ flex: .15, flexDirection: 'row', justifyContent: 'space-between', alignItems: 'flex-start', backgroundColor: '#FFF' }]} >

                    <ExitToHome navigation={this.props.navigation} color='#000'/>

                    <View style={{justifyContent: 'center', alignItems: 'center', flex: 3, height: '100%'}}>
                        <PaymentModeSwitcher charge={true} dark={true} />
                    </View>

                    <ShareWallet transferAccount={transferAccount} color='#000'/>

                </View>

                <PaymentsKeypad navigation={this.props.navigation}/>

                {/*<BottomAsyncButton buttonText={strings('NavBar.chargeButtonText').toUpperCase()} buttonColor={(transferDirection === 'charge' ? null : '#FBADA1')} onPress={(transferDirection === 'charge' ? () => this.openPaymentOptions() : () => this.openRefund())} />*/}

            </View>
        )
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(PaymentsAmountScreen);
