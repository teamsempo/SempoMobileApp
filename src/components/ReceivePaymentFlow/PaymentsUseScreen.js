import React from 'react'
import {Alert, StyleSheet, Text, View, ScrollView} from 'react-native'
import {connect} from "react-redux";
import I18n from 'react-native-i18n';

import { strings } from '../../../locales/i18n';

import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { updateTransferData } from "../../reducers/creditTransferReducer";

import TransferAmount from './TransferAmountDisplay'
import Styles from "../../Styles";
import BottomAsyncButton from "../BottomAsyncButton";

const mapStateToProps = (state) => {
    return {
        login: state.login
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        updateTransferData: (payload) => dispatch(updateTransferData(payload)),
    };
};

class PaymentsUseScreen extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            usagesSelected: [],
        }
    }

    static navigationOptions = ({ navigation }) => ({
        title: `${navigation.state.params.title}`,
    });

    _handleNext() {
        if (this.state.usagesSelected.length < 1) {
            Alert.alert(
                `${strings('PaymentsUseScreen.PaymentUse')}`,
                `${strings('PaymentsUseScreen.PaymentUsePrompt')}`,
                [{ text: strings('SendPaymentCameraScreen.Cancel'), onPress: () => console.log('cancel'), style: 'cancel'}],
            );
        } else {
            this.props.updateTransferData({transfer_use: this.state.usagesSelected});
            this.props.navigation.navigate('PaymentOption', { title: strings('GoBackNav.DefaultTitleText')});
        }
    }

    _selectPaymentUse(id) {
        console.log("id ticked", id)
        let { usagesSelected } = this.state;
        let newUsages = [...usagesSelected];

        var index = usagesSelected.indexOf(id);
        if (index !== -1) {
            newUsages.splice(index, 1);
            console.log('remove',newUsages)
            this.setState({usagesSelected: newUsages})
        } else {
            this.setState({ usagesSelected: [...usagesSelected, id]})
        }
    }

    render() {
        const { usagesSelected } = this.state;
        const { navigation } = this.props;

        if (this.props.login.transferUsages) {
            var uses = this.props.login.transferUsages.map(use_dict => {

                let useText = use_dict.name;

                let locale = I18n.currentLocale().toString()

                if (use_dict.translations) {
                    useText = use_dict.translations[locale] || use_dict.name;
                }

                let icon = use_dict.icon || 'star';


                return (
                    <View style={styles.paymentUse} key={use_dict.id}>
                        <Icon.Button key={use_dict.id} iconStyle={{marginLeft: 10, padding: 15, backgroundColor: (usagesSelected.indexOf(use_dict.id) !== -1 ? '#30a4a6' : '#fff')}} name={icon} size={20} color={usagesSelected.indexOf(use_dict.id) !== -1 ? '#FFF' : '#3D454C'}
                                     backgroundColor="#FFF"
                                     onPress={() => {
                                         this._selectPaymentUse(use_dict.id)
                                     }}>
                            <Text style={{paddingLeft: 10, width: '100%'}}>{useText}</Text>
                        </Icon.Button>
                    </View>
                )
            });
        } else {
            uses = (
                <View style={styles.paymentUse} key={0}>
                    <Icon.Button key={0}
                                 iconStyle={{
                                     marginLeft: 10,
                                     padding: 15,
                                     backgroundColor: (usagesSelected.indexOf(0) !== -1 ? '#30a4a6' : '#fff')}}
                                 name="star"
                                 size={20}
                                 color={usagesSelected.indexOf(0) !== -1 ? '#FFF' : '#3D454C'}
                                 backgroundColor="#FFF"
                                 onPress={() => {
                                     this._selectPaymentUse(0)
                                 }}>
                        <Text style={{paddingLeft: 10, paddingBottom: 10, width: '100%'}}>Other</Text>
                    </Icon.Button>
                </View>
            )
        }

        return (
            <View style={styles.rootContainer}>
                <View style={styles.mainContainer}>
                    <TransferAmount />
                    <Text style={{marginTop: 10}}>{strings('PaymentsUseScreen.PaymentUse')}</Text>
                </View>
                <View style={Styles.inputContainer}>
                    <ScrollView>
                        { uses }
                    </ScrollView>
                </View>
                <View style={Styles.inputButton}>
                    <BottomAsyncButton
                        buttonText={strings('SendPaymentCameraScreen.Next').toUpperCase()}
                        isLoading={false}
                        onPress={() => this._handleNext()}
                    />
                </View>
            </View>
        )
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(PaymentsUseScreen);


const styles = StyleSheet.create({
    rootContainer: {
        flex: 1,
        backgroundColor: '#FFF',
    },
    mainContainer: {
        flex: 5,
        justifyContent: 'center',
        alignItems: 'center',
    },
    paymentUseContainer: {
        flex: 6,
        margin: 20,
        borderTopWidth: 1,
        borderColor: '#C2C7CC',
    },
    paymentUse: {
        flexDirection: 'row',
        alignItems: 'center',
        borderBottomWidth: 1,
        borderColor: '#C2C7CC',
    },
    inputValueText: {
        color: '#3D454C',
        fontSize: 38,
        fontWeight: 'bold',
    }
});
