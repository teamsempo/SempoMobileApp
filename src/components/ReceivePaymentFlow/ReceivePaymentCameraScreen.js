'use strict';
import React, { Component } from 'react';
import {
    ActivityIndicator,
    Dimensions,
    StyleSheet,
    Text,
    View,
    Modal,
    TouchableOpacity,
} from 'react-native';
import { RNCamera } from 'react-native-camera';

import { createTransferRequest, RESET_NEW_TRANSFER } from '../../reducers/creditTransferReducer'

import { connect } from "react-redux";
import { strings } from '../../../locales/i18n';

const mapStateToProps = (state) => {
    return {
        newTransferStatus: state.creditTransfers.createStatus,
        transferData: state.creditTransfers.transferData
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        createTransferRequest: (body) => dispatch(createTransferRequest({body})),
        resetNewTransfer: () => dispatch({type: RESET_NEW_TRANSFER})
    };
};

class ReceivePaymentCameraScreen extends Component {
    constructor(props) {
        super(props);
        this.state = {
            QRCode: '',
            amount: '',
            checking_qr: false,
            invalid_qr: false,
            insufficient_funds: false,
            prompt: false,
        }
    }

    static navigationOptions = ({ navigation }) => ({
        title: `${navigation.state.params.title}`,
    });

    componentDidMount() {
        const { navigation } = this.props;
        this.props.resetNewTransfer();
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            checking_qr: (nextProps.newTransferStatus.isRequesting),
            invalid_qr: (nextProps.newTransferStatus.error === 'Incorrect Pin'),
            insufficient_funds: (nextProps.newTransferStatus.error ===  'Insufficient funds')
        });
    }

    _handleCheckQrCode(qr_id) {
        this.props.navigation.navigate('EnterPin')
    }

    _handleTransferWithDynamicQRCode(qr_data) {
        this.props.createTransferRequest({
            'qr_data': qr_data,
            'is_sending': this.props.transferData.is_sending,
            'transfer_use': this.props.transferData.is_sending? 'Refund' : this.props.transferData.transfer_use,
            'transfer_amount': this.props.transferData.transfer_amount
        });
    }

    _handleScan = (e) => {
        var data = e.data;
        if (this.state.QRCode !== data && data) {
            this.setState({
                QRCode: data
            });

            if (data.indexOf('-') === -1) {
                // This is a static QR code, for example from an NFC sticker, or it's a not a refund
                this._handleCheckQrCode(data);
            } else {
                this._handleTransferWithDynamicQRCode(data)
            }

        }
    };

    render() {
        const { height, width } = Dimensions.get('window');
        const maskRowHeight = Math.round((height - 300) / 20);
        const maskColWidth = (width - 300) / 2;

        const prompt = (this.props.transferData.is_sending? strings('ReceivePaymentCameraScreen.RefundPrompt') : strings('ReceivePaymentCameraScreen.PayPrompt'));

        if (this.state.invalid_qr) {
            var error_message = `${strings('ReceivePaymentCameraScreen.InvalidQr')}`
        } else if (this.state.insufficient_funds) {
            error_message = `${strings('ReceivePaymentCameraScreen.InsufficientFunds')}`
        } else if (this.props.newTransferStatus.error) {
            error_message = this.props.newTransferStatus.error.message
        } else {
            error_message = null
        }

        if (this.state.checking_qr) {
            var modal =
                <Modal
                    transparent={true}
                    animationType={'none'}
                    onRequestClose={() => {console.log('close modal')}}>
                    <View style={styles.modalBackground}>
                        <View style={styles.activityIndicatorWrapper}>
                            <ActivityIndicator />
                        </View>
                    </View>
                </Modal>
        } else {
            modal = <View/>;
        }

        return (
            <View style={styles.rootContainer}>
                <TouchableOpacity activeOpacity={1} style={styles.container1} onPress={() => this.setState({prompt: true})}>
                    <RNCamera
                        ref={ref => {
                            this.camera = ref;
                        }}
                        style={styles.preview}
                        type={RNCamera.Constants.Type.back}
                        flashMode={RNCamera.Constants.FlashMode.off}
                        autoFocus={RNCamera.Constants.AutoFocus.on}
                        onBarCodeRead={(e) => this._handleScan(e)}
                        barCodeTypes={[RNCamera.Constants.BarCodeType.qr]}
                        androidCameraPermissionOptions={{
                            title: `${strings('CameraScreen.permissionDialog')}`,
                            message: `${strings('CameraScreen.permissionDialog')}`,
                            buttonPositive: 'Ok',
                            buttonNegative: 'Cancel',
                          }}
                        >
                        <View style={styles.maskOutter}>
                            <View style={[{ flex: maskRowHeight  }, styles.maskRow, styles.maskFrame]} />
                            <View style={[{ flex: 30 }, styles.maskCenter]}>
                                <View style={[{ width: maskColWidth }, styles.maskFrame]} />
                                <View style={[styles.maskInner, {borderColor: (this.props.transferData.is_sending? '#FBADA1' : '#2D9EA0')}]} >
                                    {modal}
                                </View>
                                <View style={[{ width: maskColWidth }, styles.maskFrame]} />
                            </View>
                            <View style={[{ flex: maskRowHeight, justifyContent: 'center'}, styles.maskRow, styles.maskFrame]}>
                                {(error_message === null)
                                    ? <Text style={{color: (this.props.transferData.is_sending ? '#FBADA1' : '#2D9EA0'), textAlign: 'center', fontSize: 20, display: (this.state.prompt === true ? 'flex' : 'none')}}>{prompt}</Text>
                                    : <Text style={{color: '#ff0000', textAlign: 'center', fontSize: 20}}>{error_message}</Text> }
                            </View>
                        </View>
                    </RNCamera>
                </TouchableOpacity>
            </View>
        );
        }
}
export default connect(mapStateToProps, mapDispatchToProps)(ReceivePaymentCameraScreen);

const styles = StyleSheet.create({
    rootContainer: {
        flex: 1,
        backgroundColor: '#FFF',
    },
    container: {
        flex: 10,
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: '#FFF',
    },
    preview: {
        flex: 11,
    },
    promptBarContainer: {
        position: 'absolute',
        left: 0,
        right: 0,
        bottom: 0,
    },
    promptBar: {
        backgroundColor: '#FFF',
        justifyContent: 'center',
        alignItems: 'center',
        padding: 20,
    },
    capture: {
        flex: 0,
        backgroundColor: '#fff',
        borderRadius: 5,
        padding: 15,
        paddingHorizontal: 20,
        alignSelf: 'center',
        margin: 20
    },
    // NEW STYLES
    container1: {
        flex: 1,
    },
    cameraView: {
        flex: 1,
        justifyContent: 'flex-start',
    },
    maskOutter: {
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        alignItems: 'center',
        justifyContent: 'space-around',
    },
    maskInner: {
        width: 300,
        backgroundColor: 'transparent',
        borderWidth: 1,
    },
    maskFrame: {
        backgroundColor: 'rgba(1,1,1,0.6)',
    },
    maskRow: {
        width: '100%',
    },
    maskCenter: {
        flexDirection: 'row'
    },
    modalBackground: {
        flex: 1,
        alignItems: 'center',
        flexDirection: 'column',
        justifyContent: 'space-around',
        backgroundColor: '#00000040'
    },
    activityIndicatorWrapper: {
        backgroundColor: '#FFFFFF',
        height: 100,
        width: 100,
        borderRadius: 10,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-around'
    },
});