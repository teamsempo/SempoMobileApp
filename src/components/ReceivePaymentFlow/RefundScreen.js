import React, { Component } from 'react';
import {
    ActivityIndicator,
    Dimensions,
    StyleSheet,
    Text,
    View,
    Modal,
    TouchableNativeFeedback,
    TouchableOpacity,
    TextInput,
    KeyboardAvoidingView,
    Image
} from 'react-native';

import {connect} from "react-redux";
import { strings } from '../../../locales/i18n';

import QrCodeGenerator from '../SendPaymentFlow/QrCodeGenerator'
import { RESET_NEW_TRANSFER } from "../../reducers/creditTransferReducer";

const mapStateToProps = (state) => {
    return {
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        resetNewTransfer: () => {dispatch({type: RESET_NEW_TRANSFER})},
    };
};

class RefundScreen extends Component {

    componentWillUnmount() {
    }

    render() {
        const { height, width } = Dimensions.get('window');
        const maskRowHeight = Math.round((height - 270) / 20);
        const maskColWidth = (width - 270) / 2;

        var data = <QrCodeGenerator duration={5000} qrCodeSize={265}/>;
        var prompt = strings('HomeScreen.PromptText');

        return (
            <View style={styles.rootContainer}>
                <View style={styles.maskOutter}>
                    <View style={[{ flex: 30 }, styles.maskCenter]}>
                        <View style={[{ width: maskColWidth }, styles.maskFrame]} />
                        <View style={styles.maskInner} >
                            {data}
                        </View>
                        <View style={[{ width: maskColWidth }, styles.maskFrame]} />
                    </View>
                    <View style={[{ flex: maskRowHeight, justifyContent: 'center'}, styles.maskRow, styles.maskFrame]}>
                        <Text style={{color: '#000', textAlign: 'center'}}>{prompt}</Text>
                    </View>
                </View>
            </View>
        )

    }
}
export default connect(mapStateToProps, mapDispatchToProps)(RefundScreen);

const styles = StyleSheet.create({
    rootContainer: {
        flex: 1,
        backgroundColor: '#FFF',
        display: 'flex',
        alignItems: 'center',
    },
    maskOutter: {
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        alignItems: 'center',
        justifyContent: 'space-around',
    },
    maskInner: {
        width: 270,
        height: 270,
        backgroundColor: 'transparent',
    },
    maskFrame: {
        backgroundColor: 'transparent',
    },
    maskRow: {
        width: '100%',
    },
    maskCenter: {
        flexDirection: 'row'
    },
    modalBackground: {
        flex: 1,
        alignItems: 'center',
        flexDirection: 'column',
        justifyContent: 'space-around',
        backgroundColor: '#00000040',
    },
});