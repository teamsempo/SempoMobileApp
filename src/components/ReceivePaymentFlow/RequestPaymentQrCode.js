import React, { Component } from 'react';
import { Animated, StyleSheet , View} from 'react-native';
import QRCode from 'react-native-qrcode-svg';
import {connect} from "react-redux";
import Pusher from 'pusher-js/react-native';

import {generateQueryString, getHostURL, getToken} from '../../utils'

import { CREATE_TRANSFER_SUCCESS } from '../../reducers/creditTransferReducer'

const mapStateToProps = (state) => {
    return {
        login: state.login,
        transferData: state.creditTransfers.transferData,
        transferAccounts: state.transferAccounts,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        newTransferSuccess: () => dispatch({type: CREATE_TRANSFER_SUCCESS})
    };
};


class RequestPaymentQrCode extends Component {
    constructor(props) {
        super(props);
        this.state = {
            pusherChannel: null
        };
    }

    async componentDidMount() {

        let pusher = new Pusher(this.props.login.pusherKey, {
            authEndpoint: await getHostURL() + '/api/v1/pusher/auth',
            auth: {
                headers: {
                    'Authorization': await getToken(),
                }
            },
            cluster: 'ap1',
            encrypted: true,
            activityTimeout: 6000
        });

        if (pusher) {
            pusher.logToConsole = true;

            const privateUserChannel = pusher.subscribe(
                'private-user-' + this.props.login.deploymentName + '-' + this.props.login.userId
            );

            this.setState(
                {pusherChannel: privateUserChannel},
                () => {this.bindChannel(this.state.pusherChannel)}
            );
        }
    }

    componentWillUnmount() {
        if (this.state.pusherChannel) {
            this.unbindChannel(this.state.pusherChannel)
        }
    }

    bindChannel(channel) {
        channel.bind('payment_confirmed', (data) => {
            if (data.transfer_random_key === this.props.transferData.transfer_random_key) {
                this.props.navigation.navigate('TransferCompleteScreen');
                this.props.newTransferSuccess()
            }
        })
    }

    unbindChannel(channel) {
        channel.unbind('payment_confirmed')
    }


    render() {
        const { login, transferAccounts, transferData } = this.props;

        const blockchainAddressType = 'ethereum:';

        let myAccount =  transferAccounts.byId[transferData.my_transfer_account_id];
        const blockchainAddress = (myAccount && myAccount.blockchain_address) || '';

        let name = (login.firstName || '').toString() + ' ' + (login.lastName || '').toString();
        let amount = Math.round(transferData.transfer_amount/100 * 1000)/1000; //allow at most 3 decimal places, and remove cents
        let transfer_use = transferData.transfer_use;
        let user_id = login.userId;
        let token_symbol = transferData.token_symbol;
        let random_key = transferData.transfer_random_key;

        const metadata = generateQueryString({
            'amount': amount,
            'n': name,
            'tu': transfer_use,
            'u': user_id,
            's': token_symbol,
            'r': random_key,
        });

        let qr_data = blockchainAddressType + blockchainAddress + metadata;
        //ethereum:0xd25562a703f734a69405E7e1Dc4dd3B5EDabf0b9?amount=1000&f=Tristan&l=Cole&u=&rk=35E6

        return (
            <View style={styles.QrContainer}>
                <QRCode
                    value={qr_data}
                    size={this.props.qrCodeSize}

                />
            </View>
        );
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(RequestPaymentQrCode);


const styles = StyleSheet.create({
    QrContainer: {
        padding: 10,
        backgroundColor: '#ffffff',
    },
});