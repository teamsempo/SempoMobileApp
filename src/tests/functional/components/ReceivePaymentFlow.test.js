import React from 'react';
import { shallow } from 'enzyme';
import PaymentsUseScreen from '../../../components/ReceivePaymentFlow/PaymentsUseScreen';
import PaymentsAmountScreen from '../../../components/ReceivePaymentFlow/PaymentsAmountScreen';
import TransferAmountDisplay from '../../../components/ReceivePaymentFlow/TransferAmountDisplay';
import TransferCompleteScreen from '../../../components/ReceivePaymentFlow/TransferCompleteScreen';
import EnterPin from '../../../components/ReceivePaymentFlow/EnterPin';
import ReceivePaymentCameraScreen from '../../../components/ReceivePaymentFlow/ReceivePaymentCameraScreen';
import RequestPaymentQrCode from '../../../components/ReceivePaymentFlow/RequestPaymentQrCode';
import RefundScreen from '../../../components/ReceivePaymentFlow/RefundScreen';

import configureMockStore from 'redux-mock-store';

describe('PaymentsUseScreen', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({});
            const component = shallow(<PaymentsUseScreen store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});

describe('RefundScreen', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({});
            const component = shallow(<RefundScreen store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});

describe('PaymentsAmountScreen', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({ creditTransfers: { transferData: [] } });
            const component = shallow(<PaymentsAmountScreen store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});

describe('TransferAmountDisplay', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({ creditTransfers: { transferData: [] } });
            const component = shallow(<TransferAmountDisplay store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});

describe('TransferCompleteScreen', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({ creditTransfers: { transferData: [] } });
            const component = shallow(<TransferCompleteScreen store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});

describe('EnterPin', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({ 
                transferData: { transfer_amount: 5, transfer_use: 'Pizza' }, 
                createStatus: 'success',
                creditTransfers: { transferData: [] }
             });
            const component = shallow(<EnterPin store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});

describe('ReceivePaymentCameraScreen', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({ 
                transferData: { transfer_amount: 5, transfer_use: 'Pizza' }, 
                createStatus: 'success',
                creditTransfers: { transferData: [] }
             });
            const component = shallow(<ReceivePaymentCameraScreen store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});

describe('ReceivePaymentCameraScreen', () => {
    describe('Rendering', () => {
        it('should match to snapshot', () => {
            const mockStore = configureMockStore({ 
                creditTransfers: { transferData: [] }
             });
            const component = shallow(<RequestPaymentQrCode store={mockStore} />);
            expect(component).toMatchSnapshot();
        });
    });
});
