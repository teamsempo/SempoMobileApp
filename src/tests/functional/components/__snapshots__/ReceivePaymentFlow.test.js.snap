// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnterPin Rendering should match to snapshot 1`] = `
<ContextProvider
  value={null}
>
  <EnterPin
    createTransferRequest={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    transferData={Array []}
  />
</ContextProvider>
`;

exports[`PaymentsAmountScreen Rendering should match to snapshot 1`] = `
<ContextProvider
  value={null}
>
  <PaymentsAmountScreen
    removeNFCCardError={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    transferData={Array []}
    updateTransferData={[Function]}
  />
</ContextProvider>
`;

exports[`PaymentsUseScreen Rendering should match to snapshot 1`] = `
<ContextProvider
  value={null}
>
  <PaymentsUseScreen
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    updateTransferData={[Function]}
  />
</ContextProvider>
`;

exports[`ReceivePaymentCameraScreen Rendering should match to snapshot 1`] = `
<ContextProvider
  value={null}
>
  <ReceivePaymentCameraScreen
    createTransferRequest={[Function]}
    resetNewTransfer={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    transferData={Array []}
  />
</ContextProvider>
`;

exports[`ReceivePaymentCameraScreen Rendering should match to snapshot 2`] = `
<ContextProvider
  value={null}
>
  <RequestPaymentQrCode
    newTransferSuccess={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    transferData={Array []}
  />
</ContextProvider>
`;

exports[`RefundScreen Rendering should match to snapshot 1`] = `
<ContextProvider
  value={null}
>
  <RefundScreen
    resetNewTransfer={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[`TransferAmountDisplay Rendering should match to snapshot 1`] = `
<ContextProvider
  value={null}
>
  <TransferAmountDisplay
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    transferData={Array []}
  />
</ContextProvider>
`;

exports[`TransferCompleteScreen Rendering should match to snapshot 1`] = `
<ContextProvider
  value={null}
>
  <TransferCompleteScreen
    resetTransferData={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    transferData={Array []}
  />
</ContextProvider>
`;
